### ISSUE TITLE
flex/bison scripts not regenerating files properly

### ISSUE THREAD

**Issue by xexyl** [ioccc-src/mkiocccentry#177]
At first I didn't think this was an issue but I think there's a problem with the scripts for flex and bison. Look at this:

```shell
$ git status
On branch updates
Your branch is up to date with 'origin/updates'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   jparse.l
	modified:   jparse.y
	modified:   sorry.tm.ca.h
```

This was after running `make parser` and note that the C files have not been updated and neither have the header files. I'm not sure if this was introduced when I modified the Makefile to not use `-D` or if it was introduced earlier but it's an obvious problem and it's also breaking a thought I had (which I was going to put into the issues that aren't really issues thread). Any idea what might be causing this? The output of `make parser` btw:

```shell
$ make parser 
rm -f jparse.tab.c jparse.tab.h
make jparse.tab.c jparse.tab.h
rm -f limit_ioccc.sh
grep -E '^#define (RULE_|MAX_|UUID_|MIN_|IOCCC_)' limit_ioccc.h | \
	    awk '{print $2 "=\"" $3 "\"" ;}' | tr -d '[a-z]()' | \
	    sed -e 's/"_/"/' -e 's/""/"/g' -e 's/^/export /' >> limit_ioccc.sh
grep -hE '^#define (.*_VERSION|TIMESTAMP_EPOCH)' version.h limit_ioccc.h | \
	    grep -v 'UUID_VERSION' | \
	    sed -e 's/^#define/export/' -e 's/ "/="/' -e 's/"[	 ].*$/"/' >> limit_ioccc.sh
if grep -q '^#define DIGRAPHS' limit_ioccc.h; then \
	    echo "export DIGRAPHS='yes'"; \
	else \
	    echo "export DIGRAPHS="; \
	fi >> limit_ioccc.sh
if grep -q '^#define TRIGRAPHS' limit_ioccc.h; then \
	    echo "export TRIGRAPHS='yes'"; \
	else \
	    echo "export TRIGRAPHS="; \
	fi >> limit_ioccc.sh
cc -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -std=gnu11 -O0 -g -pedantic -Wall -Wextra -Werror -Wno-unused-command-line-argument -Wno-poison-system-directories -Wno-unreachable-code-break -Wno-padded -lm dyn_array.c -c
cc -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -std=gnu11 -O0 -g -pedantic -Wall -Wextra -Werror -Wno-unused-command-line-argument -Wno-poison-system-directories -Wno-unreachable-code-break -Wno-padded -lm verge.c dbg.o util.o dyn_array.o -o verge
./run_bison.sh -B /opt/homebrew/opt/bison/bin -B /opt/homebrew/bin -B /opt/local/bin -B /usr/local/opt -B /usr/local/bin -B . -p jparse -v 1 -- -d
./run_bison.sh: debug[1]: args to pass to bison before jparse.y: -d
./run_bison.sh: debug[3]: checking for bison as: bison
./run_bison.sh: debug[1]: found a good bison: /opt/local/bin/bison
/opt/local/bin/bison -d jparse.y
# prepending comment and line number reset to jparse.tab.c
cat sorry.tm.ca.h > /var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.tab.c.XXXXXXXX.m3PqLqyB
cat jparse.tab.c >> /var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.tab.c.XXXXXXXX.m3PqLqyB
mv -v -f /var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.tab.c.XXXXXXXX.m3PqLqyB jparse.tab.c
/var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.tab.c.XXXXXXXX.m3PqLqyB -> jparse.tab.c
# completed update of jparse.tab.c
# prepending comment and line number reset to jparse.tab.h
cat sorry.tm.ca.h > /var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.tab.h.XXXXXXXX.QNJvzzF4
cat jparse.tab.h >> /var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.tab.h.XXXXXXXX.QNJvzzF4
mv -v -f /var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.tab.h.XXXXXXXX.QNJvzzF4 jparse.tab.h
/var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.tab.h.XXXXXXXX.QNJvzzF4 -> jparse.tab.h
# completed update of jparse.tab.h
./run_bison.sh: debug[1]: -rw-------  1 cody  staff  69558 Apr 25 05:37 jparse.tab.c
./run_bison.sh: debug[1]: -rw-------  1 cody  staff  5446 Apr 25 05:37 jparse.tab.h
./run_bison.sh: debug[1]: bison run OK: formed jparse.tab.c and jparse.tab.h
make[1]: `jparse.tab.h' is up to date.
rm -f jparse.c
make jparse.c
./run_flex.sh -F /opt/homebrew/opt/flex/bin -F /opt/homebrew/bin -F /opt/local/bin -F /usr/local/opt -F /usr/local/bin -F . -p jparse -v 1 -- -d -8 -o jparse.c
./run_flex.sh: debug[1]: args to pass to flex before jparse.y: -d -8 -o jparse.c
./run_flex.sh: debug[3]: checking for flex as: flex
./run_flex.sh: debug[1]: found a good flex: /opt/local/bin/flex
/opt/local/bin/flex -d -8 -o jparse.c jparse.l
# prepending comment and line number reset to jparse.c
cat sorry.tm.ca.h > /var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.c.XXXXXXXX.t1CpxXrK
cat jparse.c >> /var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.c.XXXXXXXX.t1CpxXrK
mv -v -f /var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.c.XXXXXXXX.t1CpxXrK jparse.c
/var/folders/6n/nb4yym2d1r50sxlz7s0nt3dr0000gn/T/jparse.c.XXXXXXXX.t1CpxXrK -> jparse.c
# completed update of jparse.c
./run_flex.sh: debug[1]: -rw-------  1 cody  staff  65491 Apr 25 05:37 jparse.c
./run_flex.sh: debug[1]: flex run OK: formed jparse.c
```

so it appears that it's being formed but it's not being updated.

**Comment by xexyl** [comment]
Doing some checks it seems the last time it worked was prior to the new scripts. Either way this is an issue and it explains why I thought some files should have been updated but were not.

**Comment by xexyl** [comment]
Ah. I figured it out.

After running `make parser` which runs those scripts one has to do `make use_ref`. That allows the compilation to proceed. Perhaps you want to update the Makefile for this?

**Comment by xexyl** [comment]
```make
# How to create jparse.tab.c and jparse.tab.h
#
# Convert jparse.y into jparse.tab.as well as jparse.tab.c via bison,
# if bison is found and has a recent enough version, otherwise
# use a pre-built reference copies stored in jparse.tab.ref.h and jparse.tab.ref.c.
#
jparse.tab.c jparse.tab.h: jparse.y jparse.h sorry.tm.ca.h run_bison.sh limit_ioccc.sh verge jparse.tab.ref.c jparse.tab.ref.h Makefile
        ./run_bison.sh ${BISON_DIRS} -p jparse -v 1 -- -d

# How to create jparse.c
#
# Convert jparse.l into jparse.c via flex,
# if flex found and has a recent enough version, otherwise
# use a pre-built reference copy stored in jparse.ref.c
#
jparse.c: jparse.l jparse.h sorry.tm.ca.h jparse.tab.h run_flex.sh limit_ioccc.sh verge jparse.ref.c Makefile
        ./run_flex.sh ${FLEX_DIRS} -p jparse -v 1 -- -d -8 -o jparse.c

```

Appears that maybe these rules have to be updated to work on the reference files as well. But I'm not sure what way is preferred so I'll let you do it. I can think of a number of possible issues to be resolved and I'm not sure what you might want.

**Comment by lcn2** [comment]
Fixed this issue with commit 5fd6ea1e05d980e673d957390b8f6db865e476e2

    Fix make parser

    The make parser rule comment says:

       Force the rebuild of the JSON parser and form reference copies of JSON parser C code.

    The "form reference copies of JSON parser C code" has been restored.

    In addition, make parser also requires jparse.tab.o and jparse.o
    to be formed (i.e., compiled) before the reference copies are updated.

    As a side effect of fixing make parser, the reference copies of JSON parser C code
    were updated.

**Comment by xexyl** [comment]
> Fixed this issue with commit [5fd6ea1](https://github.com/ioccc-src/mkiocccentry/commit/5fd6ea1e05d980e673d957390b8f6db865e476e2)
> 
> ```
> Fix make parser
> 
> The make parser rule comment says:
> 
>    Force the rebuild of the JSON parser and form reference copies of JSON parser C code.
> 
> The "form reference copies of JSON parser C code" has been restored.
> 
> In addition, make parser also requires jparse.tab.o and jparse.o
> to be formed (i.e., compiled) before the reference copies are updated.
> 
> As a side effect of fixing make parser, the reference copies of JSON parser C code
> were updated.
> ```

Thank you! I will look at it tomorrow.
